### ISSUE TITLE
libxml2.treeError: xmlSetProp() failed when building mate-utils in Debian

### ISSUE THREAD

**Issue by lskovlun** [itstool/itstool#36]
Hi,

We've been seeing intermittent failures when building the mate-utils package in Debian, which uses itstool in its build script:

lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
lars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
llars@starfish:~/mate-utils-1.20.2/gsearchtool/help/pt$ itstool -m pt.mo ../C/index.docbook ../C/legal.xml
Traceback (most recent call last):
  File "/usr/bin/itstool", line 1598, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "/usr/bin/itstool", line 986, in merge_translations
    lcnode.setProp(attr, origlang)
  File "/usr/lib/python2.7/dist-packages/libxml2.py", line 3588, in setProp
    if ret is None:raise treeError('xmlSetProp() failed')
libxml2.treeError: xmlSetProp() failed

Instrumenting libxml2 and running itstool under the debugger yields this nugget:

(gdb) run
Starting program: /usr/bin/python /usr/bin/itstool -m pt.mo ../C/index.docbook ../C/legal.xml
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
xmlSetProp failed: 0x555555b4a300 lang en

Breakpoint 1, libxml_xmlSetProp (self=<optimized out>, args=<optimized out>) at libxml2-py.c:7281
7281            c_retval = xmlSetProp(node, name, value);
(gdb) s
xmlSetProp__internal_alias (node=node@entry=0x555555b4a300, name=0x7ffff6286264 "lang", value=0x7ffff6327784 "en") at tree.c:6831
6831        if ((node == NULL) || (name == NULL) || (node->type != XML_ELEMENT_NODE))
(gdb) print node->type
$1 = XML_TEXT_NODE
(gdb) cont
Continuing.
libxml_xmlNodePtrWrap returned None
Traceback (most recent call last):
  File "/usr/bin/itstool", line 1598, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "/usr/bin/itstool", line 986, in merge_translations
    lcnode.setProp(attr, origlang)
  File "/usr/lib/python2.7/dist-packages/libxml2.py", line 3588, in setProp
    if ret is None:raise treeError('xmlSetProp() failed')
libxml2.treeError: xmlSetProp() failed
[Inferior 1 (process 30122) exited with code 01]

As you can see, there's an assertion on the value of node->type that fails (the instrumentation repeats the failing call after printing an error message for ease of debugging). I'm not familiar with the libxml2 API, so I'm afraid I can't debug much more on my own. I'll be happy to help in any way of course.

The current Debian version of itstool is 2.0.5.

**Comment by lskovlun** [comment]
Relevant files here
[itstool-issue-36.zip](https://github.com/itstool/itstool/files/2800549/itstool-issue-36.zip)

**Comment by lskovlun** [comment]
[Debian bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=919058)
[Mate bug report](https://github.com/mate-desktop/mate-utils/issues/211)

As you can see from these bug reports, there was an initial hypothesis that this had to do with multiple parallel processes. My initial comment shows that the problem also occurs when running a single process from the command line.

**Comment by lskovlun** [comment]
It seems that moving the files out of the original directory structure and trying from there fixes the problem, so the zip file I've uploaded may not be helpful. I'd still appreciate any advice on how to debug this.

**Comment by lskovlun** [comment]
I've also instrumented the code to emit xpath expressions now, and it can be seen that the error is highly irregular even when it does occur. The number of setProp calls before the failure varies too:

libxml_xmlNodePtrWrap failed after 49 calls to libxml_xmlSetProp
/article/sect1[3]/sect2[2]/table/tgroup/thead/row/entry[2]/para/text()
libxml_xmlNodePtrWrap failed after 58 calls to libxml_xmlSetProp
/article/sect1[3]/sect2[2]/table/tgroup/thead/row/entry[2]/para/text()
libxml_xmlNodePtrWrap failed after 12 calls to libxml_xmlSetProp
/article/sect1[3]/sect2[2]/table/tgroup/thead/row/entry[2]/para/text()
libxml_xmlNodePtrWrap failed after 12 calls to libxml_xmlSetProp
/article/sect1[3]/sect2[2]/orderedlist/listitem[4]/para/guibutton/text()
libxml_xmlNodePtrWrap failed after 3 calls to libxml_xmlSetProp
/article/articleinfo/revhistory/revision[3]/revdescription/para[1]/@lang/text()
libxml_xmlNodePtrWrap failed after 66 calls to libxml_xmlSetProp
/article/articleinfo/authorgroup/author[3]/firstname/text()
libxml_xmlNodePtrWrap failed after 59 calls to libxml_xmlSetProp
/article/sect1[3]/sect2[2]/table/tgroup/tbody/row[6]/entry[1]/para/text()

**Comment by lskovlun** [comment]
Bisected. d3adf0264ee2b6fd28b7eff7dec33501d6e75a7c is the guilty commit.

**Comment by raveit65** [comment]
Hi , i got this issues with several packages in fedora 29/30, not only mate-utils.
looks like this happens randomly.

**Comment by lskovlun** [comment]
Yes, it's a use-after-free issue, introduced by the commit I indicated above. Experimentally, the copyNode fix suggested in pull request #18 works, but I'm not sure if this reintroduces the memory problems that commit was intended to fix.

**Comment by raveit65** [comment]
I happened again today https://bugzilla.redhat.com/show_bug.cgi?id=1677838

**Comment by lskovlun** [comment]
As I've noted previously, this is a use-after-free bug. But the referenced bug above (mateweather #388) indicates that the triggering factor is non-closed tags. Which makes sense, I guess - non-closed tags would require data structures to be preserved in anticipation of the missing tag, but in fact they are discarded when switching from one input file to the next (?). And the outcome is non-deterministic because of Python's memory manager. 

All of this boils down to an impedance mismatch between the automatic memory management in Python and the manual memory management required by the libxml2 wrapper. But since the proximate cause is a non-closed tag, the simplest course of action might be to add a check for non-closed tags before freeing the data structures in question. Can we do that?

**Comment by shaunix** [comment]
Hi folks. I'm aiming for a new release in the first quarter of 2021. I'd like to get all crashers like this addressed. This is a bizarre heisenbug that I can only reproduce when I stick in print statements in various places, possibly just giving a chance for freed memory to get overwritten. If anyone has tips on making this more reproducible, I'd appreciate. Having to deal with memory management in Python is not ideal.

**Comment by hvdijk** [comment]
I was hitting this myself in mate-utils more often than not, and found that running python on valgrind made it almost 100% reproducible and possible to figure out what went wrong. Created a PR that should fix this, feedback welcome, particularly if you find a system where it does not fully fix the problem.
